### PR TITLE
First version

### PR DESCRIPTION
**Pull request by pbrisbin**
[Rendered README](https://github.com/freckle/wai-middleware-openapi/blob/pb/first/README.md).

### CODE REVIEW

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (z0isch) [inline src/Network/Wai/Middleware/OpenApi.hs]: Having to parse (I know this doesn't parse but we do when we validate) the request twice (and once for the response) for all requests scares me a bit. Should we build in sampling to this library?

File: src/Network/Wai/Middleware/OpenApi/Validate.hs
Comment (z0isch) [inline src/Network/Wai/Middleware/OpenApi/Validate.hs]: Could we do something here for a missing part of the schema?

File: src/Network/Wai/Middleware/OpenApi/Validate.hs
Comment (z0isch) [inline src/Network/Wai/Middleware/OpenApi/Validate.hs]: It's nice that `openapi` has this validator built in

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (z0isch) [inline src/Network/Wai/Middleware/OpenApi.hs]: I wonder if we should support the client configuring how the error response is handled?

File: tests/TestApp.hs
Comment (z0isch) [inline tests/TestApp.hs:91]: Is there no corresponding `declareRequestBodyRef` to use?

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi.hs]: > Should we build in sampling to this library?

Yes.

Should we do it... before merging this PR? Before we release a version? Before we ourselves use it anywhere? Before we ourselves use it in a busy API? I actually don't really have an opinion; can you tell me?

File: src/Network/Wai/Middleware/OpenApi/Validate.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/Validate.hs]: I don't know what you mean by "a missing part of the schema?". Can you expand?

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi.hs]: > how the error response is handled?

Not sure what you mean by this.

File: tests/TestApp.hs
Comment (pbrisbin) [reply tests/TestApp.hs:91]: I'm not totally sure I follow. As far as I understand it, the "declare" functions all just add to a running list of declarations (`DeclareT` is just a writer). Which you can then do what you want with. I add them to `components.schemas` and then just `Ref` to them from wherever. I'm not sure what a `declareRequestBodyRef` would do or how I'd use it here.

File: src/Network/Wai/Middleware/OpenApi/Validate.hs
Comment (z0isch) [reply src/Network/Wai/Middleware/OpenApi/Validate.hs]: Doesn't a `Nothing` here imply that we hit a route that isn't in the spec?

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (z0isch) [reply src/Network/Wai/Middleware/OpenApi.hs]: I just mean a user of this library may not want to return a json error

File: tests/TestApp.hs
Comment (z0isch) [reply tests/TestApp.hs:91]: I'm mainly talking about mitigating using the stringly `OpenApi.Ref (OpenApi.Reference "Test")` below, which I thought was the point of having this `DeclareT` guy.

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (z0isch) [reply src/Network/Wai/Middleware/OpenApi.hs]: I'm really not sure either- my question was more if that's something this library *should* handle or if you saw that as something that a user of this lib would manage.

File: tests/TestApp.hs
Comment (pbrisbin) [reply tests/TestApp.hs:91]: Oh yeah, I noticed that too, but it's not actually related to declarations. It's confusing though...

I picture it something like this:

```yaml
schema:
  Test: # the actual schema for Test

requestBodies:
  Test: Ref "Test"

paths:
  "/tests":
    post:
      requestBody: Ref "Test"
```

The use of `declareSchemaRef` and `DeclareT` only handles schemas:

```yaml
# declarations = do
#   testRef <- declareSchemaRef

schemas: {declarations}

requestBodies:
  Test: {testRef}

paths:
  "/tests":
    post:
      requestBody: Ref "Test"
```

This removes one magic `"Test"` string (the name of the schema itself), but it's unrelated to the choice of the name "Test" for the request-body, which is just a simple magic string (not a schema name).

I could DRY that if we're worried, but it wouldn't be part of any declarations:

```yaml
# declarations = do
#   testRef <- declareSchemaRef
#   let testsBodyRefName = "Test"

schemas: {declarations}

requestBodies:
  {testsBodyRefName}: {testRef}

paths:
  "/tests":
    post:
      requestBody: Ref {testsBodyRefName}
```

But I think that'd be a bit silly.

You might think we could in-line schemas to solve it:

```yaml
# declarations = do
#   testRef <- declareSchemaRef

schemas: []

requestBodies: {declarations}

paths:
  "/tests":
    post:
      requestBody: {testRef}
```

But that doesn't really work for a few reasons: the `requestBodies` element is not the same type as `schemas`, and if I use `DeclareT` to declare request bodies, I can't use it to declare schemas for responses. I haven't explored that approach fully though, to be fair.

File: src/Network/Wai/Middleware/OpenApi/Validate.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/Validate.hs]: Oh yes. Sorry, I was confused when you said "a missing part of the schema"; thinking you meant a missing part of specifically the `Schema` being referenced on this line, and not "a missing part of the _spec_".

That does seem like something a strict validation shouldn't silently pass along -- but how should it behave? Does behavior depend on if the entire route is missing, vs the route being present but missing the method or response status definition?

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi.hs]: The only case I can think of where that would be useful, is if the app already has a schema defined for 400 responses and they want to make sure our own errors fit it. Can you think of any other reasons our current behavior wouldn't be appropriate?

If not, I think I'm comfortable (for now) saying that such an edge-case should be handled by dropping down to `validate`; then they can set `respondErrors` to do whatever they like. We could even ship some `defaultValidateOptions` values for them to build on. WDYT?

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi.hs]: Well that's a good question. Managing it from the outside would be more work than us managing it internally, **but** it's also an entirely independent concern.

```hs
sampledMiddleware :: Int -> Middleware -> Middleware
sampledMiddleware percent m app request respond = do
  enabled <- (<= percent) <$> randomRs (0, 100)
  if enabled
    then m app request respond
    else app request respond
```

Seems like this should exist somewhere somehow, then we don't have to do it here.

File: tests/TestApp.hs
Comment (z0isch) [reply tests/TestApp.hs:91]: Thanks for that write up! It's weird to have this nice system for declaring reusable schemas but then not be able to use it for reusable request bodies.

I'm not worried about this in the test at all, but do you even need to declare `requestBodies`? It's valid to just inline that schema right into the `requestBody`, right?

Just using a string for the schema might make some sense here too IMO. Although I guess you get to share some types between schema generation and encoding/decoding with this technique though.

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (z0isch) [reply src/Network/Wai/Middleware/OpenApi.hs]: I think a middleware wrapper like `sampledMiddleware` makes a ton of sense here, nice idea!

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (z0isch) [reply src/Network/Wai/Middleware/OpenApi.hs]: > The only case I can think of where that would be useful, is if the app already has a schema defined for 400 responses and they want to make sure our own errors fit it. Can you think of any other reasons our current behavior wouldn't be appropriate?

By schema do you mean one for the json? There's that, but it also seems reasonable to send back an html response too.

>If not, I think I'm comfortable (for now) saying that such an edge-case should be handled by dropping down to validate; then they can set respondErrors to do whatever they like. We could even ship some defaultValidateOptions values for them to build on. WDYT?

I don't think it's worth worrying about too much especially if there's a way that someone *could* do it.

File: tests/TestApp.hs
Comment (pbrisbin) [reply tests/TestApp.hs:91]: > you get to share some types between schema generation and encoding/decoding with this technique though

Yeah, this in particular was what I find valuable. It neatly avoided some false-failures in the test when I got my strings wrong.

> It's valid to just inline that schema right into the `requestBody`, right?

I think it's valid, just not sure what it'd look like to actually do. I don't think I ever have a handle to the actual schema I've declared. I have a reference to it, and I have a list of declarations to put somewhere.

I guess I could just `toSchema Proxy @NewTest`, then set that inline. But the trouble then is that if I don't declare it and set the declarations to `schemas` then any sub-schemas that are references won't be present. I don't know that that matters now, since we're just a record of primitives, but I have seen `openapi3` use references for things like `Text`, I think.

All this to say, it's complicated and this works and I'm disinclined to touch it :upside_down_face:

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi.hs]: > but it also seems reasonable to send back an html response too

Well, my thinking on that is that this middleware only operates when the request/response is `application/json`. So the only time it would be reasonable to send back HTML is if someone sent an (invalid) request at `content-type: application/json`, but was looking for a response with `accept: text/html`. That seems unlikely to me.

File: src/Network/Wai/Middleware/OpenApi/Validate.hs
Comment (z0isch) [reply src/Network/Wai/Middleware/OpenApi/Validate.hs]: > Does behavior depend on if the entire route is missing, vs the route being present but missing the method or response status definition?

I don't think that distinction really matters too much but if you can easily distinguish those then it'd be nice for the error to have that detail in it.

Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/Validate.hs]: I guess what I was getting at is the distinction may matter for what we'd want our default behavior to be. Just because a route lacks a 503 response, does that mean we (by default) want to replace it with our own "500 not in spec" error? Or what if the spec is only there for `application/json`, even though the API does have HTML responses too -- a reasonable thing to do, but if we don't add explicit checks, we'd now make those errors... and we now have to deal with content-type in our own errors... It's not sounding fun.

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi.hs]: I'm going to go ahead and document the performance concern in the README and show this example middleware. We can decide later if we should upstream such a thing to `wai-extra` (where I think it'd belong).

File: src/Network/Wai/Middleware/OpenApi/Validate.hs
Comment (chris-martin) [inline src/Network/Wai/Middleware/OpenApi/Validate.hs]:
```suggestion
-- | Internal DRY'ing of request-vs-response validation
```

File: src/Network/Wai/Middleware/OpenApi/Validate.hs
Comment (chris-martin) [reply src/Network/Wai/Middleware/OpenApi/Validate.hs]: The behavior when the route isn't in the openapi spec at all should probably be configurable in settings; I think that could be a valuable tool toward getting gradual spec coverage.

1. Start writing routes, turn on validation only for the routes that are defined in the spec so far.

2. We think we've covered everything we want to cover, change the middleware settings to reject uncovered routes. If there are routes we explicitly don't want covered, conditional application of the middleware can be done by the application outside of this library.

File: src/Network/Wai/Middleware/OpenApi/ValidationError.hs
Comment (chris-martin) [inline src/Network/Wai/Middleware/OpenApi/ValidationError.hs:46]: Is it sort of ironic that this response probably violates your openapi spec?

File: src/Network/Wai/Middleware/OpenApi/Validate.hs
Comment (chris-martin) [reply src/Network/Wai/Middleware/OpenApi/Validate.hs]: For more obscure things that a spec might be missing, e.g. schema for 503 status,

- is it fair to say that 100% OpenAPI coverage is a goal that an application can/should be working toward?
- Or is there a set of things that OpenAPI _can_ express but that an application may legitimately never care to describe?

Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/Validate.hs]: > behavior when the route isn't in the openapi spec at all should probably be configurable in settings

I think that's right. I started sketching a simple `strictMode: true`, but the issue is that once you do that, you can no longer assume `application/json`. The content-type requested, for which a schema was not found, could be anything at all. This means when we generate error responses, we now need to be _further_ configurable, or attempt to match the `content-type` of the response we we're validating / respect the `accept` header of the request we were validating. The complexity spirals really quickly.

> Or is there a set of things that OpenAPI can express but that an application may legitimately never care to describe?

I believe this is the case. I'm not going to say OpenAPI is _only useful_ for programmatic clients, but I'd say it is _mostly used_ for programmatic clients. Therefore, in the common case, it's entirely legitimate to elect not to describe anything that won't be experienced by such clients. Fancy-api has various non-JSON responses (auth, error pages, health-check, version), I see no value in spec'ing them. There are also responses generated by 3rd-party middleware. Even Servant apps, with fully generated OpenAPI, would lack those without extra work.

File: src/Network/Wai/Middleware/OpenApi/ValidationError.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/ValidationError.hs:46]: My original plan was to derive `ToSchema Error` and export it, then users of this library could add it to their own `OpenApi`. This would make our own responses valid according to ourselves.

I think I can still do that, but I realized that users might already have 400 responses specified throughout, so I can't exactly picture yet how to incorporate this schema. I ended up implicitly leaning on the "well then drop down to `validate` and change `respondError` if this matters" as the solution.

File: src/Network/Wai/Middleware/OpenApi/Validate.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/Validate.hs]: I'm going to experiment with a different user-facing interface to see if it makes solving this problem easier. It's been a long path to get to what I have here currently, and I wouldn't say I love it.

Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/Validate.hs]: OK, the latest commit should make this better. I think the interface is simpler in general, and it should make it easier for users who want to be strict to implement things exactly as they might want. I tried to speak to it a bit in the new _Configuring_ section of the README.

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (chris-martin) [inline src/Network/Wai/Middleware/OpenApi.hs]: I am guessing that the `ByteString` and `Value` are the unparsed and parsed request bodies, respectively, and the `String` is a JSON parse error? Haddocks would be cool

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (chris-martin) [inline src/Network/Wai/Middleware/OpenApi.hs:82]:
```suggestion
validateRequests spec onErrors app request respond = do
  result <- runValidateT request $ do
```

?

File: src/Network/Wai/Middleware/OpenApi/PathMap.hs
Comment (chris-martin) [inline src/Network/Wai/Middleware/OpenApi/PathMap.hs]: [Eq has a transitivity law](https://hackage.haskell.org/package/base-4.20.0.1/docs/Data-Eq.html#t:Eq) that's violated where

- `x = Exact something`
- `y = ParameterValue`
- `z = Exact somethingElse`

File: src/Network/Wai/Middleware/OpenApi/PathMap.hs
Comment (chris-martin) [inline src/Network/Wai/Middleware/OpenApi/PathMap.hs]: With `TemplatedPathComponent` not respecting `Eq`/`Ord` laws, I'm having a hard time convincing myself that this map lookup is guaranteed to do what you expect it'll do.

File: src/Network/Wai/Middleware/OpenApi/PathMap.hs
Comment (chris-martin) [inline src/Network/Wai/Middleware/OpenApi/PathMap.hs]: I'd be inclined to give `TemplatedPathComponent` a stock-derived `Eq` and `Ord`, and instead define this as a recursive structure that we do lookups from by walking down the tree one path component at a time.

```suggestion
data PathMap = PathMap
  { root :: Maybe PathItem
  , subpaths :: Map TemplatedPathComponent PathMap
  }
```

Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs]: Hmm, seems neat but I just lifted this from `wai-middleware-validation` and I don't really see anything wrong enough to do such a rewrite.

Patches welcome? If you want, feel free to commit this this PR -- though I do wish I had unit tests in place on this module (specifically `PathMap.lookup`) first to ensure a safe refactor.

Comment (chris-martin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs]: Hm, maybe will try. I don't *know* that anything is _wrong_, I am just not sure that it is _right_.

Comment (chris-martin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs]: I have been doing some testing. I'm not sure what the semantics are supposed to be here, but the present behavior smells funky.

1. If there is only a wildcard pattern `/{}`, then a lookup for path `/b` will match it.
2. If there are patterns `/a` and `/{}`, then a lookup for `/b` will match the wildcard pattern.
3. If the pattern order is flipped to `/{}` and `/a`, then a lookup for `/b` will return `Nothing`.

Surely example 3 can't be right?

Comment (chris-martin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs]: The doc section [Path Templating Matching](https://swagger.io/specification/#path-templating-matching) attempts to give some rules by illustration.

> Assuming the following paths, the concrete definition, /pets/mine, will be matched first if used:
>
> ```
> /pets/{petId}
> /pets/mine
> ```

This isn't very precisely stated, but it seems to be saying that the template with greater specificity should win, which is more or less what I would have guessed.

It goes on to say

> The following may lead to ambiguous resolution:
> 
> ```
> /{entity}/me
> /books/{id}
> ```

Which seems to just be saying that OpenAPI doesn't care to address the difficult cases and simply suggests "don't make a spec like that."

I don't see anything yet here to suggest that the order in which templates are listed is meant to influence which path matches, so I'm increasingly convinced that case 3 in the previous comment describes a bug.

Comment (chris-martin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs]: Stepping away from the code for a minute, it just hit me that what's wrong is actually really simple. When you put `/pets/{petId}` and `/pets/mine` into a `Map`, since `ParameterValue` and `Exact "mine"` compare equal, they're effectively the same map key; one gets overwritten by the other.

Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs]: Yeah, so that being wrong depends on overlapping paths semantics. Since these are ultimately `fromList`ed into the `Map`, the current implementation's semantics are implicitly "last wins". That's indeed "wrong", since the specification says it should be "most specific wins".

I don't think this is a very difficult bug to fix, or necessarily a reason to abandon the `Map` design: if you gave `TemplatedPathComponent` `<>` such that `Exact` always trumps `ParameterValue`, then `fromListWith (<>)` would prefer more specific paths, making the lookup semantics correct.

Is that "better" than your nested tree idea? Who knows.

File: src/Network/Wai/Middleware/OpenApi.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi.hs]: I always find Haddocks in this case awkward.

```hs
  -- | The request was not JSON
  --
  -- The first component of this type is the raw body and the second
  -- is the parse errors.
| RequestIsNotJson BSL.ByteString String
```

Blech.

I wish partial records weren't problematic.

```hs
  | RequestIsNotJson {rawBody :: ByteString, errors :: String}
```

would be perfectly clear.

Anyway, I'll add something.

File: src/Network/Wai/Middleware/OpenApi/PathMap.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs]: Yeah, fair enough.

File: src/Network/Wai/Middleware/OpenApi/PathMap.hs
Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs]: Well, I went down this hole a bit and it feels quite difficult to do in a Totally Correct way that is also somewhat efficient. I look forward to seeing what you come up with.

I agree the bad `Eq` is compelling, but I think as long as the tests cover all the behaviors, we can be satisfied it works correctly and then decide if it's worth stomaching to get a `Map`'s efficiency (I can't really think of a way to implement fast lookup without relying a `Map`, and so `Eq`, but maybe there is another option).

Comment (chris-martin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs]: I spent a while thinking about data structures and algorithms before figuring out that my fundamental problem is still not knowing what the correct behavior is. The way the spec is worded here,

> When matching URLs, concrete (non-templated) paths would be matched before their templated counterparts.

It seems like what somebody had in mind at one point was a simple division between

1. Static paths, which is a simple map lookup you do first
2. Templated paths, which is a wild west in the face of ambiguity

And I think this is stupid for two reasons.

First it means the "concrete path wins" property wouldn't hold up under nesting. They say that `/pets/mine` should win over `/pets/{petId}`. Okay that's fine, but if that example is following from a rule that's simply based on "is the path templated or not" then if we nest this resource under another one:

- `/pet-stores/{petStoreId}/pets/mine`
- `/pet-stores/{petStoreId}/pets/{petId}`

Now the spec seems to be saying that the routes have become ambiguous because they are both templated. Even if I read the spec that way, I cannot believe that implementations behave that way. I have to assume that the "mine" path should still take precedence.

Expanding the text's binary notion of "concrete" or "templated" into a spectrum from (more concrete, less templated) to (least concrete, most templated) begs us to invent a definition for that spectrum. Something like counting the number of wildcards in a path jumps to mind, but I don't think it lines up with what any server really does, so that seems like a pointless avenue.

Moreover I sort of resent the spec's implication that if any template could match, we are expected to find it. Realistically you should be able to expect that a server will take a left-to-right approach without backtracking.

EXAMPLE: If my spec routes are

1. `/book/{bookId}/cover`
2. `/book/{bookId}/text`
3. `/{resource}/{resourceId}/text`
4. `/{resource}/{resourceId}/ratings`

I don't like this API, but if that's really the screwed up spec that I am working with, then I should be able to reasonably assume that

- `/book/5/text` will match template 2;
- `/article/5/text` will match template 3;
- `/book/5/ratings` will 404, because the client is clearly trying to request something about a book, and books don't have ratings. In other words, once I've gone down the `book` path, I'm not backtracking to the more general thing.

The spec however says that 

- `/book/5/text` is ambiguous because it can match two templated paths; this is fine for me to disagree with because the spec leaving something ambiguous means I'm free to define a rule for it.
- `/book/5/resource-type` should match template 4; this is a pain in the ass, because it means I have to backtrack looking for a match that no sensible API would define and expect a client to be able to use in this way.

I'm going forward toying with my nested-map strategy that walks the path left-to-right. I think that when either an exact or wildcard could match a segment, you follow the exact first, and whether you backtrack or fail should be a user setting. Servers that know they have sane pattern overlaps can disable backtracking to get lookup performance that is something like O(number of path segments * log(number of path specs)). Servers that need the full path matching can enable the backtracking, and I think it'll perform fine in the common case anyway, in the worst case probably degrading to O(number of path specs).

That's all pretty easy to do, and if it really is inefficient in some way then at least it'll serve as a reference implementation to property-test something super-fast on larger specs.

Comment (chris-martin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs]: Since the purpose of this package is giving you information to help improve your server, emitting a "you should fix your path spec overlapping" warning/error on conditions like

- Multiple specs matched this path
- The spec that matched this path was only found after backtracking
 
would probably be a good feature.

Comment (pbrisbin) [reply src/Network/Wai/Middleware/OpenApi/PathMap.hs]: So I went looking at other implementations of this. Unfortunately, the only one where I could actually find this logic within their source tree is perhaps a little _too_ simple?

```js
function _pathMatcherInternal(routes, path, exactMatch) {
    return Object
        .keys(routes)
        .find((route) => {
            const routeArr = route.split('/');
            const pathArr = path.split('/');

            if (routeArr.length !== pathArr.length) return false;

            return routeArr.every((seg, idx) => {
                if (seg === pathArr[idx]) return true;

                if (!exactMatch) {
                    // if current path segment is param
                    if (seg.startsWith(':') && (idx in pathArr)) return true;
                }

                return false;
            });
        });
}

function pathMatcher(routes, path) {
    return _pathMatcherInternal(routes, path, true) || _pathMatcherInternal(routes, path, false);
}
```

The process makes sense, and this is effectively how I did it in the `text-assets/build` validation that needed a similar thing. I might add a suite of the test cases we've been discussing and see how it behaves.

**Comment by pbrisbin** [comment]
@chris-martin I think this is ready to merge, if not release. Check out the `PathMapSpec` to see how it behaves. The implementation is naive, but I think its performance scales with the number of distinct paths in the spec, which I'd expect to be manageable in even very complicated of APIs. I suspect the performance of validation overall is dominated by looking up the schema once a path is found. Still, we can always benchmark and improve later.

**Comment by chris-martin** [comment]
Starting with the straightforward linear search approach seems like a wise way to go.

**Review by chris-martin — APPROVED** [review]

**Pull request merged**
